File: problems/P-141-linked-list-cycle/solution.cpp
/*
 * LeetCode Problem: 141. Linked List Cycle
 * Difficulty: Easy
 * Topics: Linked List, Two Pointers, Hash Table
 * Date Solved: 19/06/2025
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
  int val;
  ListNode* next;
  ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
  bool hasCycle(ListNode* head) {
    if (head == nullptr)
      return false;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;

      if (slow == fast)
        return true;
    }

    return false;
  }
};

File: problems/P-141-linked-list-cycle/solution.java
/*
 * LeetCode Problem: [Problem Number]. [Problem Title]
 * Difficulty: [Easy/Medium/Hard]
 * Topics: [List of topics]
 * Date Solved: [Date]
*/

import java.util.HashSet;

class ListNode {
  int val;
  ListNode next;

  ListNode(int x) {
    val = x;
  }
}

public class solution {
  public boolean hasCycle(ListNode head) {
    if (head == null)
      return false;

    HashSet<ListNode> set = new HashSet<>();

    while (head != null) {
      if (set.contains(head))
        return true;
      set.add(head);
      head = head.next;
    }

    return false;
  }
}
